### PR TITLE
Add `useRegisterRef` hook

### PR DESCRIPTION
**Pull request by ThaNarie**
Note: the API has changed, the array registration now requires to pass the `index` as a 2nd argument.

### CODE REVIEW

**Review by strikkerpt — APPROVED** [review]

**Review by ReneDrie — APPROVED** [review]

File: src/useRegisterRef/useRegisterRef.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { Meta } from '@storybook/addon-docs';
+
+<Meta
+  title="useRegisterRef/docs"
+/>
+
+# useRegisterRef
+
+A helper hook that allows you to easily register and access refs by "name", even supports arrays.
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.stories.mdx:9]:
```suggestion
A helper hook that allows you to easily register and access refs by "name", including ref collections. 
```

File: src/useRegisterRef/useRegisterRef.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+Due to the dynamic nature of collecting refs, it's tricky to know how many `useRef` calls you need
+to consistently do without explicitly setting (and updating when your refs change) the amount.
+This is why this hook doesn't use `useRef` at all, but stores the actual DOM elements directly.
+
+This means that other hooks that rely on `ref.current` will not work. In that case you should
+change the hook (if you can), or wrap your element inside a ref (TODO: this can be easily done
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.stories.mdx:16]: Spotted a TODO ;)

File: src/useRegisterRef/useRegisterRef.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+function arrayTrimEnd<T extends Array<unknown>>(array: T) {
+  const existingElementIndex = [...array].reverse().findIndex((val) => Boolean(val));
+  const lastExistingItemIndex =
+    existingElementIndex === -1 ? 0 : array.length - existingElementIndex;
+  array.splice(lastExistingItemIndex);
+}
+
+/**
+ * A helper hook that allows you to easily register and access refs by "name", even supports arrays.
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.ts:34]:
```suggestion
 * A helper hook that allows you to easily register and access refs by "name", including ref collections.
```

File: src/useRegisterRef/useRegisterRef.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+## Usage
+
+```tsx
+const [refs, registerRef] = useRegisterRef<Refs>();
+````
+
+```tsx
+type Refs = {
+  element: HTMLDivElement;
+  elements: Array<HTMLSpanElement>;
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.stories.mdx:49]:
```suggestion
  elements: ReadonlyArray<HTMLSpanElement>;
```

File: src/useRegisterRef/useRegisterRef.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  title: 'useRegisterRef',
+};
+
+type DemoComponentProps = {};
+
+type Refs = {
+  item1: HTMLElement;
+  item2: HTMLElement;
+  btnMore: HTMLButtonElement;
+  listItems: Array<HTMLElement>;
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.stories.tsx:16]:
```suggestion
  listItems: ReadonlyArray<HTMLElement>;
```

File: src/useRegisterRef/useRegisterRef.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+  keyList: Array<HTMLElement>;
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.stories.tsx:17]:
```suggestion
  keyList: ReadonlyArray<HTMLElement>;
```

File: src/useRegisterRef/useRegisterRef.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * Usage:
+ * ```
+ *  // this gives auto-completion when registering or accessing the refs
+ *  type RefMap = {
+ *    element: HTMLDivElement;
+ *    elements: Array<HTMLElement>;
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.ts:41]:
```suggestion
 *    elements: ReadonlyArray<HTMLElement>;
```

File: src/useRegisterRef/useRegisterRef.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    index?: number,
+  ): (ref: Exclude<T[N], undefined>) => void {
+    return (ref: Exclude<T[N], undefined>) => {
+      setRefs((oldRefs) => {
+        // array mode
+        if (name.endsWith('[]')) {
+          const cleanName = name.replace('[]', '') as N;
+          if (!oldRefs[cleanName]) oldRefs[cleanName] = [] as T[N];
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const slot = oldRefs[cleanName] as Array<any>;
```
Comment (jesse-mm) [inline src/useRegisterRef/useRegisterRef.ts:85]:
```suggestion
          const slot = oldRefs[cleanName] as Array<unknown>;
```

**Pull request merged**